Add HDR output manual page (#11548)
godotengine/godot-docs: tutorials/rendering/hdr_output.rst

.. _doc_hdr_output:

HDR output
==========

HDR output is a feature that enables presentation of High Dynamic Range (HDR) visuals on HDR-capable screens. HDR **output** is not to be confused with the internal HDR rendering that is used by Godot for both Standard Dynamic Range (SDR) output and HDR output modes.

Enabling HDR output in your project
-----------------------------------

You can enable HDR output in any new or existing project using these steps:

1. Ensure *no* Environment resources use SDR-only features:

- Tonemap Mode: Filmic or ACES
- Glow Blend Mode: Soft Light
- Adjustments: Color Correction

2. Configure the :ref:`Rendering Device Driver<class_ProjectSettings_property_rendering/rendering_device/driver>` project setting to the following:

- macOS: metal
- iOS: metal

> - Windows: d3d12

Reviewer: ```suggestion
2. Configure the :ref:`Rendering Device Driver<class_ProjectSettings_property_rendering/rendering_device/driver>` project setting to the following. This is the default on projects created in Godot 4.6 onwards, but you'll need to set the Windows rendering driver manually if the project was created on an older version:

- macOS: metal
- iOS: metal
- Windows: d3d12
```

Author: I'll push a change to this soon: I found that this extra text has made this setup section a bit more difficult to read, so I added a note that follows these steps instead:

> Some of these settings may already be configured correctly for HDR output in your project. For example, the Windows Rendering Device Driver is set to ``d3d12`` in projects created in Godot 4.6 onwards, but will need to be changed if the project was created with an older version of Godot.

This way, the suggested information is still communicated, but it doesn't crowd these setup steps.

	@export_range(0, 20, 0.1, "or_less", "or_greater") var max_linear_value_limit: float = -1.0

	func _process(_delta: float) -> void:
		# Adjust the brightness of color to be the brightest possible, regardless
		# of SDR or HDR output, but no brighter than max_linear_value_limit.
		var max_linear_value = get_window().get_output_max_linear_value()
		if max_linear_value_limit >= 0.0:
			max_linear_value = minf(max_linear_value, max_linear_value_limit)
		self_modulate = normalize_color(sdr_self_modulate, max_linear_value)

	func normalize_color(srgb_color, max_linear_value = 1.0):
		# Color must be linear-encoded to use math operations.
		var linear_color = srgb_color.srgb_to_linear()
		var max_rgb_value = maxf(linear_color.r, maxf(linear_color.g, linear_color.b))
		var brightness_scale = max_linear_value / max_rgb_value
		linear_color *= brightness_scale
		# Undo changes to the alpha channel, which should not be modified.
		linear_color.a = srgb_color.a
		# Convert back to nonlinear sRGB encoding, which is required for Color in
		# Godot unless stated otherwise.
		return linear_color.linear_to_srgb()

Using a ``WorldEnvironment`` node
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> To produce HDR output using a WorldEnvironment, your scenes will need Color values that exceed what an SDR screen can present, so it is important to use a tonemapper like Reinhard or AgX to handle display of bright scene values on both SDR and HDR screens. 

Reviewer 2: ```suggestion
To produce HDR output using a WorldEnvironment, your scenes will need color values that exceed what an SDR screen can present, so it is important to use a tonemapper like Reinhard or AgX to handle displaying bright scene values on both SDR and HDR screens.
```

Not all screens are equal
--------------------------

SDR standards were designed to match the capabilities of existing screens that were commonly used around the world. HDR standards have been intentionally written with the opposite approach: they are designed to utilize the capabilities of an ideal screen that is not yet widely available.

In practice, this means that common HDR screens may perform their own internal tonemapping to support content that extends to a wider gamut and luminance range than what the physical hardware can achieve. This tonemapping may produce colors that are not representative of other screens, so it's best to disable this feature, if possible, when developing your HDR game. You may be able to disable some or all internal tonemapping by enabling the HGiG mode on your screen or setting the screen's mode to "clip" and/or "stable".

> Additionally, some HDR screens may present dark or saturated colors differently than others. This difference in appearance is often the result of the screen and device technologies. Devices with built-in HDR screens often have the most accurate HDR presentation of dark or saturated colors, but this may change over time as technologies develop.

Author: The text suggested here is good, though, so I'm curious if there is a precedent for providing Godot users with shopping/technology advice in the manual like this and if it's a good idea and maintainable in the long term.

I'm also curious about OLED (not QD-OLED), like I have in my iPhone 13 Pro. It doesn't suffer from the same raised black levels from ambient light issue. And the mini-LED on my Macbook Pro is by far my most capable and versatile HDR screen, especially due to the lack of any internal tonemapping, so I would be quick to recommend this device/screen for game development, but I don't see how a user would come to that realization from reading the proposed text...

Enabling HDR output in your project
-----------------------------------

You can enable HDR output in any new or existing project using these steps:

1. Ensure *no* Environment resources use SDR-only features:

- Tonemap Mode: Filmic or ACES
- Glow Blend Mode: Soft Light
- Adjustments: Color Correction

2. Configure the :ref:`Rendering Device Driver<class_ProjectSettings_property_rendering/rendering_device/driver>` project setting to the following:

- macOS: metal
- iOS: metal
- Windows: d3d12

3. Turn on the :ref:`HDR 2D<class_ProjectSettings_property_rendering/viewport/hdr_2d>` project setting and enable :ref:`use_hdr_2d<class_Viewport_property_use_hdr_2d>` of all :ref:`SubViewports <class_SubViewport>` and :ref:`Windows <class_Window>` that should support HDR output.
4. Turn on the [code]display/window/hdr/request_hdr_output[/code] project setting and enable :!ref:`hdr_output_requested<class_Window_property_hdr_output_requested>` of all other :ref:`Windows <class_Window>` that should support HDR output.

> 5. [Optional] Provide in-game settings for the player to adjust HDR brightness and maximum luminance by copying [code]in_game_hdr_settings[/code] from this demo into your project.

Reviewer 2: ```suggestion
5. [Optional] Provide in-game settings for the player to adjust HDR brightness and maximum luminance by
   copying ``in_game_hdr_settings`` from this demo to your project.
```

HDR output fundamentals
-----------------------

Godot uses the `Extended Dynamic Range (EDR) <https://developer.apple.com/videos/play/wwdc2022/10113/>`__ paradigm for HDR output. While Standard Dynamic Range (SDR) output allows color component values between 0.0 and 1.0 to be displayed, HDR output allows values higher than 1.0. The maximum value that can be displayed is provided by Window.output_max_linear_value and is valid when using SDR or HDR.

(Image showing [0, infinity] with [0, 1 aka output_max_linear_value for SDR] and [0, ~3 aka output_max_linear_value for HDR]. Maybe four bars, greyscale, red, green, blue in the centre with SDR on the left and HDR on the right.)

Designing for HDR output
------------------------

There are two primary approaches to make the most of HDR output: using Window.output_max_linear_value and using the WorldEnvironment. While both approaches can be used in the same project, the output_max_linear_value should not be used in scenes that are affected by a WorldEnvironment.

Using output_max_linear_value
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> In a traditional SDR-only game, the brightest presentation of a color is limited by either the red, green, or blue component of the color reaching a maximum of 1.0. Using modern HDR screens, this limitation no longer applies and color components above 1.0 can be accurately presented by an HDR screen. Godot provides the maximum colour component value that can be presented by the screen through the Window.output_max_linear_value property. This property can be used in both SDR and HDR, which makes it easy to build your game for both output modes without needing to change bevhaior based on whether or not HDR is enabled.

Reviewer 2: ```suggestion
In a traditional SDR-only game, the brightest presentation of a color is limited by either the red, green, or blue component of the color reaching a maximum of 1.0. Using modern HDR screens, this limitation no longer applies and color components above 1.0 can be accurately presented by an HDR screen. Godot provides the maximum color component value that can be presented by the screen through the Window.output_max_linear_value property. This property can be used in both SDR and HDR, which makes it easy to build your game for both output modes without needing to change behavior based on whether or not HDR is enabled.
```
Also references and line lengths

It is best to use this output_max_linear_value with "highlights" and special effects that are either brief or involve a small portion of the screen; if the majority of the screen is presented at this maximum brightness for more than a short time, it will cause the game to appear uncomfortably bright, as if the game is ignoring the device brightness setting. You may also find that some effects should be limited to a maximum linear value that is greater than 1.0, but less than output_max_linear_value.

Transforming a colour to be the brightest the screen can present can be done with a script. When working with CanvasItems, it may be convenient to apply the resulting modified color to the self_modulate or modulate property with the base color of the CanvasItem set to Color.WHITE. The following script demonstrates this:

> ```

Reviewer 2: Needs to use the
```
.. tabs::
 .. code-tab:: gdscript GDScript
```
Syntax

.. _doc_hdr_output:

HDR output
==========

HDR output is a feature that enables presentation of High Dynamic Range (HDR) visuals on 
HDR-capable screens. HDR **output** is not to be confused with the internal HDR rendering that 
is used by Godot for both Standard Dynamic Range (SDR) output and HDR output modes.

Enabling HDR output in your project
-----------------------------------

You can enable HDR output in any new or existing project using these steps:

1. Ensure *no* :ref:`Environment<class_environment>` resources use SDR-only features:

- Tonemap Mode: Filmic or ACES
- Glow Blend Mode: Soft Light
- Adjustments: Color Correction

2. Configure the :ref:`Rendering Device 
   Driver<class_ProjectSettings_property_rendering/rendering_device/driver>` project setting to 
   the following. This is the default on projects created in Godot 4.6 onwards, but you'll need to 
   set the Windows rendering driver manually if the project was created on an older version:

- macOS: metal
- iOS: metal

> - Windows: d3d12

Reviewer 3: We should probably mention that Linux will need to use the Wayland Display Server (`display_server/driver.linuxbsd`) as well. Wayland is also required in the editor to see HDR there, which is a different setting (`run/platforms/linuxbsd/prefer_wayland`).

:ref:`tonemap white<class_Environment_property_tonemap_white>` will be shown without clipping. 
This ensures that details are not lost when presenting the image on a screen with a lower 
dynamic range than the original scene.

.. image:: img/rendering_hdr_output_tonemap_white.jpg

While this behavior is perfectly stable in SDR, where the :ref:`output max linear 
value<class_Window_method_get_output_max_linear_value>` is fixed at ``1.0``, this behavior is 
dynamic with HDR based on the capabilities of the screen:

.. image:: img/rendering_hdr_output_hdr_tonemap.webp

As shown in the graphs above, the 
:ref:`Reinhard<class_Environment_constant_tone_mapper_reinhardt>` tonemapper will behave the 
same as the :ref:`Linear<class_Environment_constant_tone_mapper_linear>` tonemapper when 
:ref:`output max linear value<class_Window_method_get_output_max_linear_value>` is equal to or 
higher than :ref:`tonemap white<class_Environment_property_tonemap_white>`. This allows for 
accurate color reproduction on HDR screens that are capable of reproducing the original 
brighter scene values. When :ref:`output max linear 

> value<class_Window_method_get_output_max_linear_value>` has increased to a be higher than 

Author: ```suggestion
value<class_Window_method_get_output_max_linear_value>` has increased to be higher than 
```

effect is directly influenced by the brightness of the scene. If the scene brightness changes 
based on :ref:`output max linear value<class_Window_method_get_output_max_linear_value>`, then 
the glow strength will change as well: a larger :ref:`output max linear 
value<class_Window_method_get_output_max_linear_value>` will produce a stronger glow effect, 
which is generally an undesirable behavior.

An exception to this rule is to change the :ref:`background 
mode<class_Environment_property_background_mode>` through a :ref:`class_WorldEnvironment` with 
a :ref:`Linear<class_Environment_constant_tone_mapper_linear>` tonemapper because the 
:ref:`Linear<class_Environment_constant_tone_mapper_linear>` tonemap mode disables tonemapping 
and the background rendering is not affected by the brightness of scene values.

Absolute luminance values
-------------------------

When using HDR output, :ref:`output max linear 
value<class_Window_method_get_output_max_linear_value>` is calculated based on the reference 
white luminance (the device brightness setting) and the maximum luminance of the screen.

Reference white luminance
^^^^^^^^^^^^^^^^^^^^^^^^^

> The reference white luminance, or reference luminance for short, is simply the device 

Reviewer 3: I'm not sure we should describe this as "simply the device brightness setting" as that doesn't seem to be true for all displays. For instance, my LG monitor is constantly as "max brightness" but still has a `output_max_linear_value` of around `5.0`.

Instead, maybe we should describe it as the brightness of SDR content on the display as configured by the OS (aka. paperwhite or SDR white point)?
